src/cdomains/valueDomain.ml

```diff
@@ -587,31 +550,37 @@ struct
       warn_type "widen" x y;
       `Top
 
+  let rec smart_join x_eval_int y_eval_int  (x:t) (y:t):t =
+    let join_elem: (t -> t -> t) = smart_join x_eval_int y_eval_int in  (* does not compile without type annotation *)
+    match (x,y) with
+    | (`Struct x, `Struct y) -> `Struct (Structs.join_with_fct join_elem x y)
+    | (`Union (f,x), `Union (g,y)) -> `Union (match UnionDomain.Field.join f g with
+        | `Lifted f -> (`Lifted f, join_elem x y) (* f = g *)
+        | x -> (x, `Top)) (* f <> g *)
+    | (`Array x, `Array y) -> `Array (CArrays.smart_join x_eval_int y_eval_int x y)
+    | (`Blob x, `Blob y) -> `Blob (Blobs.join x y)
```

Comment:
The blob case looks the same, the smart version just had an extra comment before.

Reply:
Nice catch!